### PR TITLE
[FIRRTL] Add Aggregate creation ops and aggregate constants

### PR DESCRIPTION
**Pull request by darthscsi**
Add support for directly making aggregates rather than having to use a wire.  Also support constant aggregates.  Add canonicalization to transform aggregate wires with field-wise writes to single writes.

### CODE REVIEW

**Comment by darthscsi** [comment]
Shouldn't be applied until lowering is ready.

**Comment by darthscsi** [comment]
TODO: check for don't touch.

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp [hunk 1]
```diff
@@ -1729,6 +1734,167 @@ void NodeOp::getCanonicalizationPatterns(RewritePatternSet &results,
...
 
+SmallVector<Value> getCompleteWrite(Operation *lhs) {
+  auto lhsTy = lhs->getResult(0).getType();
+  if (!lhsTy.isa<BundleType>() && !lhsTy.isa<FVectorType>())
+    return {};
+
+  DenseMap<uint32_t, Value> fields;
+  for (Operation *user : lhs->getResult(0).getUsers()) {
+    if (user->getParentOp() != lhs->getParentOp())
+      return {};
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/FIRRTLFolds.cpp:1745]: Is this possible in FIRRTL Dialect? Or is this specifically avoiding the case involving assignments in `when`s?

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp [hunk 2]
```diff
@@ -1729,6 +1734,167 @@ void NodeOp::getCanonicalizationPatterns(RewritePatternSet &results,
...
+    if (auto aConnect = dyn_cast<StrictConnectOp>(user)) {
+      if (aConnect.getDest() == lhs->getResult(0))
+        return {};
+    } else if (auto subField = dyn_cast<SubfieldOp>(user)) {
+      for (Operation *subuser : subField.getResult().getUsers()) {
+        if (auto aConnect = dyn_cast<StrictConnectOp>(subuser)) {
+          if (aConnect.getDest() == subField) {
+            if (fields.count(subField.getFieldIndex())) // duplicate write
+              return {};
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/FIRRTLFolds.cpp:1754]: Nice! Thanks for checking for multiple assignment. I was worried that this canonicalizer would be a foot-gun if it did the wrong thing before `ExpandWhens`. 👍

**Review by seldridge — COMMENTED** [review]
This is looking great! This is going to really compact some code coming out of Chisel.

**Comment by seldridge** [comment]
Also, how much of this could be done via ODS? 😉

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp [hunk 3]
```diff
@@ -1729,6 +1734,167 @@ void NodeOp::getCanonicalizationPatterns(RewritePatternSet &results,
   results.insert<FoldNodeName>(context);
 }
 
+SmallVector<Value> getCompleteWrite(Operation *lhs) {
```
[the remark this answers is not in the document]
Comment (darthscsi) [reply lib/Dialect/FIRRTL/FIRRTLFolds.cpp:1737]: no, return by value is better and preferred.

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp (hunk 1 above)
Comment (darthscsi) [reply lib/Dialect/FIRRTL/FIRRTLFolds.cpp:1745]: yes, it's to not touch things prior to when elimination.

**Comment by darthscsi** [comment]
> Also, how much of this could be done via ODS

none.  ODS basically falls down for variadic things.  It is also highly biased towards data-flow patterns, which make connect-rooted optimizations hard to deal with.

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp [hunk 4]
```diff
@@ -1723,12 +1729,180 @@ struct NodeBypass : public mlir::RewritePattern {
...
+    } else if (auto subField = dyn_cast<SubfieldOp>(user)) {
+      for (Operation *subuser : subField.getResult().getUsers()) {
+        if (auto aConnect = dyn_cast<StrictConnectOp>(subuser)) {
+          if (aConnect.getDest() == subField) {
+            if (fields.count(subField.getFieldIndex())) // duplicate write
+              return {};
+            fields[subField.getFieldIndex()] = aConnect.getSrc();
+            continue;
+          }
+        }
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/FIRRTLFolds.cpp:1760]: Rejecting all users except for connections might be too conservative. I believe we can allow most of operations which use the subfield as source. 
```
wire array: UInt<1>[2];
array[0] <= 0;
array[1] <= 0;
node a = array[0];
````
I agree that the current implementation is sufficient to clean up temporary wires for aggregate constants so
can we add TODO comment to improve that in the future?

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/FIRRTLFolds.cpp]: Why is it better/preferred, would you mind elaborating?  More often than I'd like I'm not sure so tend to be explicit about the behavior I want which isn't always great, especially if the simpler and clearer version is handled at least as well (and I'd rather not fight the compiler!).

I could be mistaken (apologies if so!), but from looking at the compiled code (clang 14.0.6, FWIW), the move-assignment operator for `SmallVector<mlir::Value>` is invoked before checking `values.empty()` (with `DenseMap` cleanup between) in the caller of `getCompleteWrite`-- which will copy the elements if still "small", is that expected/right?  Not sure it's a big deal but unsure why that's better (vs populating caller's copy directly).  This is a Debug+Asserts build so maybe that pessimizes things? Regardless of this particular instance, general guidelines always good to know! Thanks!

**Pull request merged**